Binary snapshots (in-memory)

lasernoises opened the pull request:
The biggest difference to the previous attempt in #489 is that the snapshot contents are now kept in-memory instead of being written into a temporary file directly.

For `SnapshotContents` I decided to make it an enum with two variants like in the previous attempt. I first attempted to change it to three variants so it would also replace the `SnapshotKind` enum but I realized that a lot of logic is still shared between inline and file text snapshots. That approach would definitely also work though and it's perhaps also a question of how much the logic between the two is expected to diverge in the future.

In `runtime::assert_snapshot` I replaced the combination of the `refval` and `new_snapshot_value` with an enum. In that case it is a three variant enum and some `From` impls for two value tuples for the macros. Maybe it would be better to make it a two variant enum there too for consistency and keep the `ReferenceValue` enum for the text case. That might be a bit weird because both binary and non-inline text snapshot have a name and that would then be in two different enums.

I should probably also add some tests in `cargo-insta`. I'll do that as soon as possible, but I need to figure out how the testing there works.

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -51,43 +55,86 @@ macro_rules! elog {
-        ReferenceValue::File(value.map(Cow::Owned))
+type SnapshotName<'a> = Option<Cow<'a, str>>;
+
+pub enum SnapshotValue<'a> {
```
Can we keep / add the docstrings from https://github.com/mitsuhiko/insta/pull/610/files#diff-ab23e8b51879793793aa0d764a47aa9084b29b94ac7b70e58a9067752ae9e781L87 ?

max-sixty commented on insta/src/snapshot.rs:
```diff
@@ -511,42 +596,107 @@ impl Snapshot {
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct BinarySnapshotContents {
```
We have a lot of new data structures! I'm thinking whether it's possible to reduce how many we need. As an example — is this needed, or can a binary snapshot just store the bytes? The only thing this seems to do is to build its path, which arguably the binary snapshot should do already?

lasernoises replied:
I don't quite know what you meant with that last part. Because a `Snapshot` doesn't know its own path, that only gets passed in from context. Which also means that we don't need to mutate it when approving for example. This also extends to the binary snapshots, which build the paths on demand based on the relevant pending or stored path. In the #489 implementation I had the path in the `SnapshotContents` and that made some things a little annoying and weird because we needed to update that in the right places and so on.

But anyway, we do already have the extension in the `MetaData` struct which is redundant. So I've changed it to use the extension from there. I suppose the whole thing is not quite optimal because we don't have the typesystem enforcing that the snapshot type in the metadata and in the contents is the same, but enforcing that would be a pretty big change I think because it would likely mean not having the `metadata` field in `Snapshot` struct, but like everything from metadata except for the exension and then generating the full metadata at the moment of saving or something like that. So I think assuming them to be the same thing plus asserting and unwrapping in some places is an acceptable trade-off.

max-sixty replied:
I very much agree with your second paragraph, I've been thinking around the same lines. There's a decent amount of duplication in the data structures, which is a smell.

Possibly there's some big refactor we could do in the future (should we have `SnapshotAssertionContextFile` given lots there is specific? `SnapshotAssertionContext<Kind>`?? How to generalize across FileText & Binary; vs FileText & InlineText?). But I don't think we have to compress everything at this point, as long as the code is nicely legible.

Thanks for trying the change in https://github.com/mitsuhiko/insta/pull/610/commits/1b4f7ec471fc50356a222bf7b24dce3c9f5b4722. After seeing it, I think it's possibly slightly better, though not as much of a simplification as I had imagined, partly because we can't specialize on the Binary vs Text at the Snapshot struct. Am fine either way if you want to keep / revert.

max-sixty commented on insta/src/snapshot.rs:
```diff
@@ -511,42 +596,107 @@ impl Snapshot {
 
     /// Snapshot matches based on the latest format.
-    pub fn matches_latest(&self, other: &SnapshotContents) -> bool {
+    pub fn matches_latest(&self, other: &TextSnapshotContents) -> bool {
```
(I should have done this before)
```suggestion
    pub fn matches_latest(&self, other: &Self) -> bool {
```

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -641,17 +697,41 @@ pub fn assert_snapshot(
+        SnapshotValue::Binary {
+            content, extension, ..
+        } => {
+            assert!(extension != "new", "this file extension is not allowed");
```
not important for now, but as we get closer, we should do things like include `new` in the error message here

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -51,43 +55,86 @@ macro_rules! elog {
+type SnapshotName<'a> = Option<Cow<'a, str>>;
+
+pub enum SnapshotValue<'a> {
+    Text {
```
To the extent we keep this enum — the naming isn't ideal, where `Text` includes `Inline` elsewhere but not here?

max-sixty reviewed (commented):
It's in impressive effort! And thank you for redoing on top of the new `master`.

It does add lots more data structures & some complications to the code. I'm going to think about how much we could simplify (I've already simplified the existing code somewhat..). If you have ideas, that would be great too.

max-sixty commented on insta/src/output.rs:
```diff
@@ -103,121 +103,183 @@ impl<'a> SnapshotPrinter<'a> {
+            (Some(SnapshotContents::Binary(_)) | None, SnapshotContents::Text(new)) => {
+                Some((None, Some(new.to_string())))
+            }
+            (Some(SnapshotContents::Text(old)), SnapshotContents::Binary { .. }) => {
```
Presumably we should never we comparing these?

(No great problem, though it also makes me think whether we can overall simplify this sort of code...)

lasernoises replied:
Well, you can change a text snapshot to a binary one and vice versa since they essentially live in the same namespace. Also I did find that feature somewhat helpful when switching some things to binary snapshots where previously I had a snapshoted a struct with measurements.

I suppose the alternatives are to either just show something like `- old text snapshot` in red or show nothing at all (like would be the case with inline snapshots, where in order to replace it you have to delete the old snapshot in the code), but I would prefer to show the old contents somehow and I think the way this displays currently is quite reasonable:

![image](https://github.com/user-attachments/assets/ece82450-34c7-4189-afdc-d98447bd2634)

But I'm definitely open to suggestions for a nicer UI for this.

max-sixty replied:
Ah yes, good point, I hadn't thought of that. 

(this would be the sort of thing that would be great to add an integration test — i.e. what happens if we switch between text and binary...)

max-sixty commented on cargo-insta/tests/main.rs:
```diff
@@ -1145,3 +1162,191 @@ fn test_hashtag_escape() {
+    +      src/snapshots/test_binary_change_extension__binary_snapshot.snap
+    +      src/snapshots/test_binary_change_extension__binary_snapshot.snap.json
+    ");
+}
```
(Nice integration tests!)

max-sixty commented on insta/src/snapshot.rs:
```diff
@@ -511,42 +596,111 @@ impl Snapshot {
         Ok(new_path)
     }
+
+    pub fn cleanup_extra_files(snapshot: Option<&Self>, path: &Path) -> Result<(), Box<dyn Error>> {
```
Why do we need this function? (I don't doubt we do, I just don't quite get grok it at first...)

lasernoises replied:
If the extension of a pending snapshot changes when re-running an unapproved test (or goes from a binary to a text snapshot) it would otherwise just write it with the new extension and overwrite the .snap. The alternative would be to try to read the pending file path of what is being asserted and then delete the file with the extension from in there. And then I also used it when approving or rejecting a pending snapshot, just because that's a bit simpler than using the extension to derive the path to delete (but in that case the metadata is already in memory).

This approach was more necessary with the other implementation where we created the file as a `._` first which also had to be cleaned up if a crash happened anywhere in between creating it at the beginning of `assert_snapshot` and where the `.snap` was actually being written.

But I still think this is the simpler approach compared to loading the previous pending snapshot and looking at the extension field there.

max-sixty replied:
Can we integrate with the existing cleanup functions — either `cleanup_passing` or removing unreferenced snapshots? 

Those file tree walkers that will find snapshots reliably too.

lasernoises replied:
If I understood this correctly, I think the unreferences snapshot removal happens fully in `cargo-insta`. I think this needs to happen when just running with `cargo test` as well, otherwise there can be a bunch of different files lying around.

To integrate it with `cleanup_passing` we'd have to change that into a general `cleanup` function that doesn't only run in the passing case. Also currently `cleanup_passing` only tries to delete one file: the one with the `.new` appended. And `cleanup_extra_files` currently relies on being run before the new files are stored, because otherwise it would delete the new pending snapshot.

I gave it a shot and it ended up looking like this:

```rust
pub fn cleanup(&self, passing: bool) -> Result<(), Box<dyn Error>> {
    if let Some(ref path) = self.snapshot_file {
        let binary_path =
            self.old_snapshot
                .as_ref()
                .and_then(|snapshot| match &snapshot.contents() {
                    SnapshotContents::Text(_) => None,
                    SnapshotContents::Binary(contents) => Some(contents.build_path(path)),
                });
        let binary_file_name = binary_path.as_ref().map(|path| path.file_name().unwrap());

        // The file name to compare against has to be valid utf-8 as it is generated by this
        // crate out of utf-8 strings.
        let file_name = path.file_name().unwrap().to_str().unwrap();

        let pending_path = path.with_extension("snap.new");
        let pending_file_name = pending_path.file_name().unwrap();

        let pending_binary_path =
            self.old_snapshot
                .as_ref()
                .and_then(|snapshot| match &snapshot.contents() {
                    SnapshotContents::Text(_) => None,
                    SnapshotContents::Binary(contents) => {
                        Some(contents.build_path(&pending_path))
                    }
                });
        let pending_binary_file_name = pending_binary_path
            .as_ref()
            .map(|path| path.file_name().unwrap());

        let read_dir = path.parent().unwrap().read_dir();

        match read_dir {
            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(()),
            _ => (),
        }

        for entry in read_dir? {
            let entry = entry?;
            let entry_file_name = entry.file_name();

            // We'll just skip over files with non-utf-8 names. The assumption being that those
            // would not have been generated by this crate.
            if entry_file_name
                .to_str()
                .map(|f| f.starts_with(file_name))
                .unwrap_or(false)
                && entry_file_name != file_name
                && binary_file_name.map_or(true, |x| x != entry_file_name)
                && (passing
                    || (entry_file_name != pending_file_name
                        && pending_binary_file_name.map_or(true, |x| x != entry_file_name)))
            {
                std::fs::remove_file(entry.path())?;
            }
        }
    }

    if passing {
        // and add a null pending snapshot to a pending snapshot file if needed
        if let Some(ref pending_snapshots) = self.pending_snapshots_path {
            if fs::metadata(pending_snapshots).is_ok() {
                PendingInlineSnapshot::new(None, None, self.assertion_line)
                    .save(pending_snapshots)?;
            }
        }
    }

    Ok(())
}
```

But then I realized that it's also used in `cargo-insta` in `Container::commit`. There we don't have an `AssertionContext` (and probably also shouldn't). So we'd need a different solution there.

Right now I can't really think of a way to integrate the two that doesn't increase the overall complexity.

max-sixty replied:
OK, I get some of this, though haven't quite groked why we need a separate path. Will have another look through later. 

If possible, I think we should try and integrate with the existing functions, so the semantics are the same and the internals remain simple. For example — while this isn't a huge deal — it looks like if we change the extension, we [remove unreferenced snapshots](https://github.com/mitsuhiko/insta/blob/654e1a6ee8038a616d254fd8376f4e199aab6780/cargo-insta/tests/main.rs#L1328) (nice tests!) — but we generally wouldn't do that (without a `--delete-unreferenced-snapshots` run)

For example, we already have the `try_removing_snapshot` on the pending path (https://github.com/mitsuhiko/insta/blob/8753f2cc59284bd8bc8c645594e524fd16495423/cargo-insta/src/container.rs#L231-L242) — could we instead extend that to also remove the a pending binary snapshot when it's accepted? (Maybe the answer to this is "not without making the code more complicated, because that just takes a path, and this would need to take an object which looked up the pending paths")

lasernoises replied:
Ok, I've changed some things around. This function was a little bit weird because it deleted all the files that start with the passed in path. While that was correct, it's a little bit confusing and I think the exact way I did it is related to there being an extra temporary file in the previous implementation.

I've now changed the function to only deal with any additional `.snap.new.*` files in with the same snapshot filename. This is the alternative to loading the previous pending snapshot and then deleting the binary snapshot from there. (We could also do that, but in that case I'd want to only load the metadata to prevent having an extra copy of the binary in-memory). In cargo-insta I've changed it to just delete the binaries there, because the snapshots are already loaded.

max-sixty replied:
Thank you!

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -51,43 +55,86 @@ macro_rules! elog {
+
+    InlineText {
+        reference_content: &'a str,
+        content: &'a str,
```
Do we ever use both of these fields? Can we add docstrings here too?

lasernoises replied:
I've added some docstrings to this enum. Regarding the first question: do you mean use both of them together? They get used separately in `assert_snapshot`. The only reason they're in the same enum is only to prevent the invalid `Inline + Binary` combination. This is basically the first variant of what I had suggested in [this comment](https://github.com/mitsuhiko/insta/pull/489#issuecomment-2168227959). But I'm now strongly considering doing something more like the second variant for more consistency (and probably less code because I could replace all those `From` impls with only one).

(thx for the quick feedback btw)

max-sixty replied:
Does the same information go in each field? I'm still a bit unclear what each field is for. (I'm sure it's not rocket science, but I'm also thinking of the next person who comes across the code and wants to understand it...)

Or ofc feel free to just do the refactor if you think that makes it simpler!

max-sixty replied:
Ah, is `content:` the generated value? If so, let's use that term in the docstrings.

max-sixty commented:
> For `SnapshotContents` I decided to make it an enum with two variants like in the previous attempt. I first attempted to change it to three variants so it would also replace the `SnapshotKind` enum but I realized that a lot of logic is still shared between inline and file text snapshots. That approach would definitely also work though and it's perhaps also a question of how much the logic between the two is expected to diverge in the future.

FWIW I think the ontology is quite an interesting case — `FileText` & `FileBinary` share some properties, but `FileText` & `InlineText` also share some completely different properties. So there's no perfect hierarchy.

The proposed structure seems reasonable (as above I'd like to see if we can reduce the number of data structures a bit, but I don't think there are any easy large clean-ups)

max-sixty commented:
To sketch out what I think we need to do to merge:
- Tests — we already have a couple good integration tests. Would be great to extend those to any corner cases — for example, what if someone uses the same extension for a text snapshot? What does it look like if someone has an older version of `cargo-insta`? (maybe those are too specific, but anything in that direction)
- @mitsuhiko needs to review, I think focused on the external API
  - Obv great if we can make the internal code nice, but also we can do that later, and would be a loss if this withered based on refinements.
- We should decide whether we mark as experimental. If we don't mark it as experimental, we're stuck with the API, and in some ways insta has more stringent backward-compatibility constraints than other libraries, since it needs to work with older versions of itself in the form of `cargo-insta`)
- Docs — we should add a short section on this in https://github.com/mitsuhiko/insta-website

max-sixty commented on cargo-insta/tests/main.rs:
```diff
@@ -81,6 +81,17 @@ fn assert_success(output: &std::process::Output) {
+    eprint!("{}", String::from_utf8_lossy(&output.stdout));
+    assert!(
+        !output.status.success(),
+        "Tests succeeded: {}\n{}",
```
```suggestion
        "Tests failed as expected: {}\n{}",
```

lasernoises replied:
This is in the case that the assert doesn't hold. But this message is indeed confusing, so I've changed it to `"Tests unexpectedly succeeded"`.

max-sixty replied:
Sorry yes, thanks

max-sixty commented:
Re additional tests: we should ensure we cover named snapshots. Possibly I'm making a mistake but this doesn't compile atm:

```rust
#[test]
fn test_binary_snapshot_named() {
    insta::assert_binary_snapshot!("txt", "name", b"test".to_vec());
}
```

---

I'm thinking of whether we can retain more of the structure of existing familiar macros, rather than using different parameters in the form of `$extension, $name, $value`. (developers of the library can undervalue the benefits of consistency because we're so "in" it, while most folks reading tests would highly value consistency; particularly for macros, which don't easily expose the names of their parameters)

Not confident, but one possible way to retain `$name, $value` is to push `$extension` into the `$name` param:
- `*.tar.gz` would be used for an unnamed snapshot
- `name.tar.gz` would be used for a named snapshot. The resulting file would be `${prefix}_name.snap.tar.gz`

I think this would be reasonably simple? On the downside, `name.tar.gz` looks like a file name, and we don't exactly get that file name; and it's not possible to name a snapshot beginning with `*`.

lasernoises commented:
You're right, the named binary snapshots were broken. I've now fixed them and also changed how the macros work in in the text case. The change is not necessarily an obvious win, because it introduces a little bit of weirdness with an ignored parameter, but I do think I makes it a bit clearer what is going on and it requires a little bit less `From` magic.

Syntax wise we have a lot of options for how to pass the extensions and I'm not quite sure what the best one is. We could possibly even use ident tokens or something like that, but I suppose that's not optimal either because it would cause weirdness around keywords.

max-sixty commented:
> What I'm a bit more worried about is the `internals` module. I am changing some things that are exposed in there. The docstring does imply that it's primarily for documentation. So that might imply that it's not part of the stable API?

Yes, this is `hidden` and so not part of the public API (somewhat by convention). We only have it public so `cargo-insta` can access it.

> I think older versions of `cargo-insta` will definitely not be able to support binary snapshots. I think that's simply not solvable.

For sure — my point re older `cargo-insta` versions is: if we release this in 1.41 and then realize we should have done it differently, we can't break the 1.41 format, unless we've marked it as experimental. And we're a bit more constrained than other apps because things like pending snapshots are an interface between `cargo-insta` & `insta`, which themselves can be on different versions...

> that would mean we'd need to parse that string at runtime because I don't think the contents of a string literal can be parsed by a `macro_rules` macro.

Yes, but I think that's fine? The existing approach for text snapshots — `Autoname` — is at runtime too, albeit the macro knows at compile time which path it'll take.

I would be less keen on the ident approach — I don't see much benefit and it's less consistent.

max-sixty commented on cargo-insta/tests/main.rs:
```diff
@@ -1350,3 +1350,54 @@ fn test_binary_snapshot() {
+         src/main.rs
+    +    src/snapshots
+    ");
+}
```
Nice test!

max-sixty commented on cargo-insta/tests/main.rs:
```diff
@@ -1401,3 +1401,165 @@ fn test_binary_snapshot() {
+    +    src/snapshots
+    +      src/snapshots/test_change_binary_to_text__some_name.snap
+    ");
+}
```
And these! (though also would be nice to explain in a docstring on the test what it's doing...)

lasernoises replied:
I've added some docstrings. Hope they make sense.

lasernoises commented on cargo-insta/src/container.rs:
```diff
@@ -213,22 +217,28 @@ impl SnapshotContainer {
             for snapshot in self.snapshots.iter() {
                 match snapshot.op {
                     Operation::Accept => {
-                        let snapshot = Snapshot::from_file(&self.pending_path).map_err(|e| {
```
I'm not sure about this. But it seemed to me that loading the snapshot again here is redundant. It seems this was introduced in #218 originally (the code was still in `cargo.rs` then). If we just use the snapshot from `snapshot.new` here it should also mitigate the problem referenced in #516. The only reason I can see for loading it again separately would be to support updating pending snapshots while reviewing, but it seems a lot of other things could go wrong there as well.

But it's totally possible, that I'm just missing something.

max-sixty replied:
Yes, very much agree, great spot.

(though I also may be missing something...)

rafaelbeckel commented on cargo-insta/src/inline.rs:
```diff
@@ -4,7 +4,7 @@ use std::fs;
 use std::path::{Path, PathBuf};
 
-use insta::_cargo_insta_support::SnapshotContents;
+use insta::_cargo_insta_support::TextSnapshotContents;
```
I might be missing some context here, but why we need to introduce a separate enum for text? Why not just create a new variant for it? I.e. `SnapshotContents::File`, `SnapshotContents::Inline`, and `SnapshotContents::Binary`?

lasernoises replied:
Basically because there is still a lot of shared behavior between file and inline snapshots. By having the `TextSnapshotContents` struct we can use that in the places where we match on the contents and call methods on it. If inline and file were separate variants with separate strings all of this would be much more annoying. You'd have to match with a `|` pattern in the places with shared behavior and then either still call methods on the main `SnapshotContents` and `unwrap` an `Option` it returns or have a freestanding functions to replace the methods that just take a string parameter.

I tried implementing it this way for a bit, but gave up because of all of those problems. It would definitely be doable, but a larger change and likely some additional duplication of code.

max-sixty commented on insta/src/snapshot.rs:
```diff
@@ -540,7 +538,10 @@ impl Snapshot {
+            (SnapshotContents::Binary(a), SnapshotContents::Binary(b)) => {
+                // For binary snapshots the extension also need to be the same:
+                a == b && self.metadata.snapshot_type == other.metadata.snapshot_type
+            }
```
tiny nit: we could just call `.matches` in `.matches_fully` and encapsulate this there

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -616,8 +699,7 @@ where
 pub fn assert_snapshot(
-    refval: ReferenceValue,
-    new_snapshot_value: &str,
+    new_snapshot_value: SnapshotValue<'_>,
```
nit: this isn't just the "new" value; it's both...

max-sixty commented on insta/src/macros.rs:
```diff
@@ -333,35 +333,73 @@ macro_rules! assert_compact_debug_snapshot {
-    (transform=$transform:expr, $name:expr, $value:expr, $debug_expr:expr $(,)?) => {
+    // The extra value gets ignored here because above we can't only pass $value if there's no
+    // $debug_expr due to macro_rules not having anything like an else.
+    ($snapshot_value:expr, $debug_expr:expr $(, $_:expr)? $(,)?) => {
```
FYI I had a go at trying to clean this up, I didn't manage. I worry a tiny bit with this sort of code (ignoring args) that we're going to miss something and break something (test coverage is improving but still not complete...).

But we have to be able to make progress, so a worthwhile tradeoff if it's necessary.

To the extent we collapse the name & extension into one, like `*.jpg`, I think we can restore much of the previous structure. If we don't, possibly there's a design that's closer to the original, such passing the kind in another arg.

max-sixty replied:
@lasernoises to the extent we can undo some of these changes now that the macros have a similar structure, that would be great... (I realize it's a bit of a bore, but I am really keen to keep this code simple, hope that's reasonable to ask...)

max-sixty replied:
Thank you!

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -228,11 +255,12 @@ struct SnapshotAssertionContext<'a> {
     assertion_file: &'a str,
     assertion_line: u32,
     is_doctest: bool,
+    snapshot_type: SnapshotType,
```
nit: more idimoatic to name this `snapshot_kind`, and will become part of the public API

max-sixty commented:
@mitsuhiko I'll DM you in the next few days about this PR. The things we want from you:
- OK in principle?
- Any views on what the macro params should be? The currently proposed code has `$extension, ($name,)? $value, ($debug_expr)?`.
    - I offered a proposal [above](https://github.com/mitsuhiko/insta/pull/610#issuecomment-2372105382) for `$name.extension, $value, ($debug_expr)?`, where `name` can be `*`.
- OK to merge when we're happy with it, after we've added docs etc? Or you want to review again?

mitsuhiko commented:
Sorry for ghosting this for so long.  These PRs always require actual brainpower to be put to it, so they tend to slip down the list automatically when I have limited capacity for it.  So first of all: definitely on board with having this.

I will give this some more deeper thoughts but here are some very immediate questions that came up.  The most obvious one is this one:

> ```rust
> assert_binary_snapshot!("some_name.jpg", b"test".to_vec());
> ```

This was rejected above there because the string requires parsing at runtime.  Is this a problem?  Because at least in theory this makes the most sense from a user experience point of view and all the other stuff proposed seems a bit odd.

lasernoises commented:
> > ```rust
> > assert_binary_snapshot!("some_name.jpg", b"test".to_vec());
> > ```
> 
> This was rejected above there because the string requires parsing at runtime. Is this a problem? Because at least in theory this makes the most sense from a user experience point of view and all the other stuff proposed seems a bit odd.

I just would have preferred to make it a compile time error if the `.` was missing. We could also interpret it just as an extension in that case, but I don't think that's a good idea.

I totally agree that all of the other options I layed out are a bit weird and not great. Also given that this is a testing library I think checking things at runtime is much less of an issue than for other libraries so I suppose the thing about not being able to check the format of the string at compile time is more of an annoyance than a real issue.

I'll give this a go tomorrow. I think for unnamed snapshots I'll just use `".jpg"` instead of `"*.jpg"` because I think that makes it look less like a glob, which I think could make it a bit confusing when reading one of those for the first time. But that might just be me.

max-sixty commented:
> . I think for unnamed snapshots I'll just use `".jpg"` instead of `"*.jpg"` because I think that makes it look less like a glob, which I think could make it a bit confusing when reading one of those for the first time.

I'd be OK with this. The downside is it prevents using a leading `.` as a file name (though _technically_ `*.jpy` prevents using a leading `*` as a file name!).

And the `*.jpy` looks sufficiently different from a name. But also understand it looks like a glob. Open to other ideas; could use a `$` or `_` or `{name}` etc.

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -98,6 +103,31 @@ pub enum SnapshotValue<'a> {
+    ) -> Self {
+        let (name, extension) = name_and_extension
+            .split_once('.')
+            .expect("the snapshot name and extension should be in the format \"name.extension\"");
```
(we should say what was passed etc)

max-sixty commented:
Great, thanks @lasernoises , I think the functionality is good!

From the list above: https://github.com/mitsuhiko/insta/pull/610#issuecomment-2369071150, we should add some docs here, I would suggest marking it as experimental. And if possible we could add a very short section on the website, and then I think we're good to go. (a couple of code clean-ups too)

(I'll let @mitsuhiko weigh-in if he has other feedback, but won't wait for another review unless he requests one...)

max-sixty commented on cargo-insta/tests/main.rs:
```diff
@@ -918,12 +923,13 @@ Hello, world!
      source: src/lib.rs
     -expression: 
     +expression: "\"Hello, world!\""
+    +snapshot_type: text
```
One tiny change before we merge is to rename this to `_kind` to be more idiomatic, given it'll become part of our interface

lasernoises commented:
> From the list above: [#610 (comment)](https://github.com/mitsuhiko/insta/pull/610#issuecomment-2369071150), we should add some docs here, I would suggest marking it as experimental. And if possible we could add a very short section on the website, and then I think we're good to go. (a couple of code clean-ups too)

I hope I'll have some time tomorrow to add something to the website.

Regarding the experimental thing: Should I just add a doc comment to the macro that states that the feature is experimental? Also: if we consider the feature experimental, maybe it would make sense to not write the `snapshot_kind` metadata field for text snapshots yet?

---

On an unrelated note: I was trying to test if the feature for deleting unreferenced snapshots works correctly for binary snapshots, but I can't seem to get that to remove unreferenced snapshots at all, even on master. It might be an issue with my machine or something like that.

From the code I have a suspicion that once I get it to work it will not handle binary snapshots right yet. So it probably makes sense to wait with merging this until I've figured out what I'm doing wrong there.

max-sixty commented:
> Regarding the experimental thing: Should I just add a doc comment to the macro that states that the feature is experimental?

Yes great! And the changelog / basically anywhere we mention it. We can say something in the macro docstring like "we may make incompatible changes for the next couple of versions after 1.41"

> maybe it would make sense to not write the `snapshot_kind` metadata field for text snapshots yet?

I think it's fine to add. I think maybe 20% chance we change something about the interface somewhere, but that piece seems quite robust.

> I was trying to test if the feature for deleting unreferenced snapshots works correctly for binary snapshots, but I can't seem to get that to remove unreferenced snapshots at all, even on master

Whoops, this no longer works — not your machine! Will fix and add a test (test coverage _gradually_ growing from a very low base!)

max-sixty commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@ All notable changes to insta and cargo-insta are documented here.
 
 - Inline snapshots only use `#` characters as delimiters when required.  #603
 
+- Experimental support for binary snapshots
```
```suggestion
- Experimental support for binary snapshots.  #610 (Florian Plattner)
```

(or feel free to change to `@lasernoises`, your call

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -623,25 +764,46 @@ pub fn assert_snapshot(
+        } => {
+            assert!(
+                extension != "new",
+                "'.new' is not allowed as a file extension"
```
```suggestion
                "'new' is not allowed as a file extension"
```

(I realize confusion about whether we start extensions with periods was part of the cause for the Unreferenced bug, so am trying to ensure we're consistent. Though open-minded if you think we should instead _always_ use a leading period...)

max-sixty commented on insta/src/macros.rs:
```diff
@@ -359,9 +359,44 @@ macro_rules! _assert_snapshot_base {
+///
+/// The contents get stored in a separate file next to the metadata file. The extension for this
+/// file must be passed as part of the name. For an implicit snapshot name just an extension can be
+/// passed starting with a `.`.
```
Great — could we add a couple of examples? This will be the main source of docs for people. And then we can basically copy & paste to the website. (Check out the other macros for examples of examples)

max-sixty replied:
Super, thanks!

max-sixty commented on insta/src/runtime.rs:
```diff
@@ -623,25 +764,46 @@ pub fn assert_snapshot(
+        } => {
+            assert!(
+                extension != "new",
+                "'new' is not allowed as a file extension"
```
```suggestion
                "'.new' is not allowed as a file extension"
```

Sorry, mea culpa, I didn't think enough, reverting my suggestion

max-sixty commented:
A couple of examples and then let's hit the big button. Thank you *very* much for all your persistence here, I know it was a long road. It's a great feature for insta!

lasernoises commented:
The deleting of unreferenced snapshots does not work for binary snapshots yet. I'm working on fixing that now.

lasernoises commented:
Now it works, but I'm not quite sure what the best error handling strategy there is. It needs to load the snapshot now to know whether it's binary and to get the extension. It still ignores errors when deleting the files like before. So I think it should probably also ignore errors when loading the snapshot and just skip it and maybe print a log message.

lasernoises commented on cargo-insta/src/cli.rs:
```diff
@@ -811,6 +844,12 @@ fn handle_unreferenced_snapshots(
             }
             eprintln!("  {}", path.display());
             if matches!(action, Action::Delete) {
+                let snapshot = Snapshot::from_file(&path)?;
```
```suggestion
                let snapshot = match Snapshot::from_file(&path) {
                    Ok(snapshot) => snapshot,
                    Err(e) => {
                        eprintln!("Error loading snapshot at {:?}: {}", &path, e);
                        continue;
                    }
                };
```

Maybe something like that?

max-sixty replied:
Yes that's great!

max-sixty commented on insta/src/snapshot.rs:
```diff
@@ -517,39 +608,59 @@ impl Snapshot {
+    // of the `Vec` could be rather large. The reason it's not an `Rc<[u8]>` is because creating one
+    // of those would require re-allocating because of the additional size needed for the reference
+    // count.
+    Binary(Rc<Vec<u8>>),
```
We don't have to resolve before merging but — are we actually cloning this much? I tried removing the `Rc`, everything seemed to work, and I couldn't immediately see where we were cloning it.

lasernoises replied:
The place I saw the clone was in `runtime::record_snapshot_duplicate`. While that doesn't happen all the time I think it's worth it because wrapping it in an `Rc` is so easy.

Just as one reference point: the PDF files I'm snapshotting ended up being around 600kb usually. That's not a huge ammount of data, but compared to bumping a reference count it is a bunch of work and I think if we can prevent that copy so easily I would lean on the side of just doing it.

max-sixty replied:
Yes, good point. And seems difficult to change given it needs to hold the value throughout the test run...

lasernoises commented:
> > What I'm a bit more worried about is the `internals` module. I am changing some things that are exposed in there. The docstring does imply that it's primarily for documentation. So that might imply that it's not part of the stable API?
> 
> Yes, this is `hidden` and so not part of the public API (somewhat by convention). We only have it public so `cargo-insta` can access it.

I meant to reply to this earlier but forgot:
I think you meant the `_cargo_insta_support` module. Because the `internals` module is actually not hidden. I think given that it's called `internals` we can probably justify changing things exposed in it. Maybe it would make sense to add a comment there that explicitly states that this is not guaranteed to be stable.

lasernoises commented:
I've started adding a section for the website: https://github.com/mitsuhiko/insta-website/compare/main...lasernoises:insta-website:binary-snapshots?expand=1. I think I'll wait with opening a PR until this is merged.

max-sixty commented:
> I think you meant the `_cargo_insta_support` module. Because the `internals` module is actually not hidden. I think given that it's called `internals` we can probably justify changing things exposed in it. Maybe it would make sense to add a comment there that explicitly states that this is not guaranteed to be stable.

Yes v good point. Would be great to add the comment (but no need to slow down this PR ofc)

max-sixty commented:
@lasernoises anything else? Or we merge? :)

lasernoises commented:
I think from my point of view we can merge.

lasernoises commented:
🎉🎉🎉

Thank you so much for your patience and the many reviews/comments!

max-sixty commented:
Thank you!!
